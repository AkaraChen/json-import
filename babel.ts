import type babel from "@babel/core";
import type { PluginObj } from "@babel/core";

export default function plugin({ types: t }: typeof babel): PluginObj {
    return {
        visitor: {
            ImportDeclaration(path, state) {
                const { node } = path;
                // only handle json import
                if (
                    !node.source.value.endsWith(".json") &&
                    !(node.assertions?.at(0)?.value?.value === "json")
                ) return;
                // `import { createProxy } from 'json-import'` if not imported
                if (!state.file.path.scope.hasBinding("createProxy")) {
                    path.insertBefore(
                        t.importDeclaration(
                            [
                                t.importSpecifier(
                                    t.identifier("createProxy"),
                                    t.identifier("createProxy")
                                ),
                            ],
                            t.stringLiteral("json-import")
                        )
                    );
                }
                // insert proxy declaration
                path.insertAfter(
                    t.variableDeclaration("const", [
                        t.variableDeclarator(
                            t.identifier(node.specifiers[0].local.name),
                            t.callExpression(
                                t.identifier("createProxy"),
                                [t.stringLiteral(node.source.value)]
                            )
                        ),
                    ])
                );
                // delete `import json from './test.json' assert { type: 'json' };`
                path.remove();
            },
        },
    };
}
