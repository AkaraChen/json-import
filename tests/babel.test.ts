import babel from '@babel/core'
import { test, expect } from 'vitest'
import plugin from '../babel'
import env from '@babel/preset-env'
import syntaxImportAssertions from '@babel/plugin-syntax-import-attributes'

test('babel', () => {
    const result = babel.transformSync(
        "import json from './test.json' with { type: 'json' };",
        {
            presets: [env],
            plugins: [plugin, syntaxImportAssertions],
        }
    )
    expect(result?.code).matchSnapshot()
})
