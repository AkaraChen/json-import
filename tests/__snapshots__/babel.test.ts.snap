// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel 1`] = `
""use strict";

var _jsonImport = require("json-import");
var json = (0, _jsonImport.createProxy)("./test.json");"
`;
